### DISCUSSION TITLE
[Chapter 07] 불변리스트 퀴즈

### DISCUSSION THREAD

**Discussion by snoopingh** [cowaine/is-my-code-that-strange#60] [Chapter 07-08 (컬렉션, 강한 결합)] [answered]
Quiz) p145에  코드7,17 외부에서 변경하지 못하도록 불변 리스트로 만들어서 전달하기 라는 내용이 있습니다

`
List<Member> members() {
    return members.unmodifiableList();
}
`

위 코드에서 unmodifiableList() 를 반환하면서 "불변 리스트로 만들어서 전달한다" 라고 설명하고 있는데,
과연 이 리스트가 정말 불변일까요?

**Comment by firewoody237** [comment, marked as answer]
오 일단 다른 스터디원분들의 질문을 읽지않고 저의 질문을 올렸는데요, 비슷한 맥락의 질문이었던 것 같습니다.

전 아래와같은 코드를 짜보면서, "완벽한 불변"은 아니라고 생각했습니다.
아래처럼 참조요소에 대해서는 보호해주지 못하더라구요.

그냥 메서드에 대해서만 safe한 녀석으로 보여요.

![image](https://github.com/cowaine/is-my-code-that-strange/assets/113076211/64427191-9c98-4da6-b2ea-d719c64d1982)

**Reply by firewoody237** [reply]
https://github.com/cowaine/is-my-code-that-strange/discussions/63

굉장히 비슷한 질문이라 제질문은 훑기만 해도 될 것 같습니다 ㅎㅎ

**Comment by corock** [comment]
참고로 코드 7.17 은 저자가 잘못 설명한 부분이 있습니다. `List` 인터페이스에는 `unmodifiableList` 메서드가 없어요.

![unmodifiable-list](https://github.com/cowaine/is-my-code-that-strange/assets/38161720/1789f291-7bbc-4e13-9c31-e6561d5d4b5a)

그래서 `Collections.unmodifiableList(members)` 또는 스트림 API 에서 `Collectors` 클래스를 활용했다고 가정하면 좋을 것 같습니다.

![collectors-to-unmodifiable-list](https://github.com/cowaine/is-my-code-that-strange/assets/38161720/55b3f2cd-cbcf-4ac0-862f-358266cb50a6)

그리고 퀴즈에 대한 답은 앞서 나온 대로 불변이 아니라 불변성을 지키지 않았을 때 `UnsupportedOperationException` 예외가 발생합니다.

**Reply by coalong** [reply]
추가로 궁금한게 UnsupportedOperationException 런타임 예외를 뱉고있는데, 불변성을 지켜야 할 상황에는 new list 로 생성하는 것이 좋지 않나요?
